File: remove_background.py
#!/usr/bin/env python3
"""
Remove white/light background from sprite sheet and make it transparent
"""
from PIL import Image
import sys

def remove_white_background(input_path, output_path, threshold=240):
    """
    Remove white/light colored background from image
    
    Args:
        input_path: Path to input image
        output_path: Path to save output image
        threshold: RGB value threshold for considering a pixel as "white"
    """
    # Open image
    img = Image.open(input_path)
    img = img.convert("RGBA")
    
    # Get pixel data
    datas = img.getdata()
    
    newData = []
    for item in datas:
        # If pixel is whitish/grayish (all RGB values above threshold)
        # make it transparent
        if item[0] > threshold and item[1] > threshold and item[2] > threshold:
            newData.append((255, 255, 255, 0))  # Transparent
        else:
            newData.append(item)  # Keep original
    
    # Update image data
    img.putdata(newData)
    
    # Save as PNG with transparency
    img.save(output_path, "PNG")
    print(f"✓ Saved transparent version to: {output_path}")

if __name__ == "__main__":
    input_file = "/Users/nox/Library/Mobile Documents/com~apple~CloudDocs/bodygogo/public/pixel_girl_spritesheet_v4.png"
    output_file = "/Users/nox/Library/Mobile Documents/com~apple~CloudDocs/bodygogo/public/pixel_girl_spritesheet_v4_transparent.png"
    
    print(f"Processing: {input_file}")
    remove_white_background(input_file, output_file, threshold=230)
    print("Done!")

File: remove_background.js
/* eslint-disable @typescript-eslint/no-require-imports */
/**
 * Remove white background while preserving skin tones
 * More conservative approach to protect character colors
 */
const sharp = require('sharp');
const path = require('path');

const INPUT_PATH = path.join(__dirname, 'public', 'pixel_girl_spritesheet_v4_original.png');
const OUTPUT_PATH = path.join(__dirname, 'public', 'pixel_girl_spritesheet_v4_original_transparent.png');

async function removeWhiteBackground() {
    try {
        console.log('Processing:', INPUT_PATH);

        const image = await sharp(INPUT_PATH)
            .ensureAlpha()
            .raw()
            .toBuffer({ resolveWithObject: true });

        const { data, info } = image;
        console.log('Image size:', info.width, 'x', info.height);

        // More conservative threshold to preserve skin tones
        // Skin tones typically have lower RGB values than pure white
        const threshold = 240; // Only remove very bright pixels
        let pixelsChanged = 0;

        for (let i = 0; i < data.length; i += 4) {
            const r = data[i];
            const g = data[i + 1];
            const b = data[i + 2];

            // Only target near-white pixels
            // Skin tones have more variation between RGB channels
            const isVeryWhite = r > threshold && g > threshold && b > threshold;

            // Additional check: RGB values should be very close to each other (white/gray)
            const colorVariance = Math.abs(r - g) + Math.abs(g - b) + Math.abs(r - b);
            const isGrayscale = colorVariance < 30; // Low variance = grayscale

            if (isVeryWhite && isGrayscale) {
                // Make completely transparent
                data[i] = 255;
                data[i + 1] = 255;
                data[i + 2] = 255;
                data[i + 3] = 0; // Alpha: fully transparent
                pixelsChanged++;
            }
            // Otherwise keep original pixel unchanged (preserves skin tones)
        }

        console.log(`Changed ${pixelsChanged} pixels to transparent`);

        // Save processed image
        await sharp(data, {
            raw: {
                width: info.width,
                height: info.height,
                channels: 4
            }
        })
            .png({ compressionLevel: 9 })
            .toFile(OUTPUT_PATH);

        console.log('✓ Saved to:', OUTPUT_PATH);
        console.log('✓ Skin tones should be preserved!');

    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

removeWhiteBackground();
